### PR TITLE
Pl20 7081

### PR DESCRIPTION
**Pull request by ldapster**

### CODE REVIEW

File: BIDECDSA.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+	pubKey, err := secp256k1.ParsePubKey(tot)
+	if err != nil {
+		log.Printf("CreateSharedKey error: %v %v", err, len(pubBytes))
+	}
+	privBytes, _ := base64.StdEncoding.DecodeString(privKeyStr)
+	privKey := secp256k1.PrivKeyFromBytes(privBytes)
+	return secp256k1.GenerateSharedSecret(privKey, pubKey)
+}
+
+func encryptGcm(aes cipher.Block, plainText string) (string, error) {
```
Comment (manik-1kosmos) [inline BIDECDSA.go]: I thought all of encrypt / decrypt should be with GCM.. @ldapster is that not the case? why do you need a dedicated "ecnryptGcm" "decryptGcm"

File: BIDECDSA.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	}
+	encryptedBytes, nonce := unpackCipherData(encrypted, aesgcm.NonceSize())
+	decryptedBytes, err := aesgcm.Open(nil, nonce, encryptedBytes, nil)
+	if err != nil {
+		return "", err
+	}
+	return string(decryptedBytes[:]), nil
+}
+
+func packCipherData(cipherText []byte, iv []byte, tagSize int) string {
```
Comment (manik-1kosmos) [inline BIDECDSA.go:120]: Is this public function? can this be made private?

File: BIDECDSA.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	ivLen := len(iv)
+	data := make([]byte, len(cipherText)+ivLen)
+	copy(data[:], iv[0:ivLen])
+	copy(data[ivLen:], cipherText)
+	return base64.StdEncoding.EncodeToString(data)
+}
+
+func unpackCipherData(data []byte, ivSize int) ([]byte, []byte) {
```
Comment (manik-1kosmos) [inline BIDECDSA.go:128]: Is this public function? can this be made private?

File: BIDECDSA.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+		return Encrypt(text, key)
+	} else if method == DECRYPT {
+		return Decrypt(text, key)
+	} else {
+		return "", fmt.Errorf("EcdsaHelper invalid method %s", method)
+	}
+}
+
+func Encrypt(plainText string, key []byte) (string, error) {
+	aes, err := aes.NewCipher(key)
```
Comment (manik-1kosmos) [inline BIDECDSA.go:53]: based on below "gcm" function, is this one doing CBC? We must only use GCM please

**Review by manik-1kosmos — CHANGES_REQUESTED** [review]
Please see comments

File: BIDECDSA.go (hunk 4 above)
Comment (ldapster) [reply BIDECDSA.go:53]: no Galois Counter Mode (GCM)

File: BIDECDSA.go (hunk 1 above)
Comment (ldapster) [reply BIDECDSA.go]: removed dedicated encryptGcm/decryptGcm

File: BIDECDSA.go (hunk 2 above)
Comment (ldapster) [reply BIDECDSA.go:120]: private function

**Review by manik-1kosmos — APPROVED** [review]

**Pull request merged**